torch/distributed/_sharded_tensor/__init__.py

```diff
@@ -360,26 +360,18 @@ def pre_load_state_dict_hook(module, state_dict, prefix, local_metadata, strict,
     _recurse_update_module(module, state_dict, prefix)
 
 def _recurse_update_module(module, state_dict, prefix):
-    for attr_name, attr in module.__dict__.items():
-        key = prefix + attr_name
-        if key in state_dict:
-            if isinstance(state_dict[key], ShardedTensor):
-                setattr(module, attr_name, state_dict[key])
-
     for submodule_name, submodule in module.named_modules():
-        key = prefix + submodule_name
-        if submodule_name:
-            _recurse_update_module(submodule, state_dict, key + '.')
-
+        for attr_name, attr in module.__dict__.items():
```

Comment:
Shouldn't this be `submodule.__dict__.items()`? I'm wondering why this test didn't catch this issue: https://github.com/pytorch/pytorch/blob/master/test/distributed/_sharded_tensor/test_sharded_tensor.py#L965? Can we enhance that test so that it does catch this issue (maybe we need multiple levels of nesting)?

Reply:
You are right, it should be submodule. 
I'm not yet sure why the test had passed, but I'll give it a try.

Reply:
@pritamdamania87 I'm a little bit stuck here. 
The test will pass no matter what. I'm running

`python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_state_dict`

I even added an `assert False` and it still passed. The output is always:

```
.
----------------------------------------------------------------------
Ran 1 test in 1.523s

OK
```

Reply:
This is indeed weird, can you try `pytest test/distributed/_sharded_tensor/test_sharded_tensor.py -k test_state_dict`?

cc @janeyx99 I was wondering if there might be something wrong here with our test infra?

Reply:
Thanks for the suggestion @pritamdamania87  
The same phenomenon happens with your command. Output looks normal, as if all tests passed. 
However, I found out that it is because of the @with_comms decorator. This one makes the test pass without running it. By removing it, my `self.assertEqual(1, 2)` triggers as expected. 

Any intuition how this decorator was meant to be used?

Reply:
@wanchaol This issue seems to be coming from the changes in https://github.com/pytorch/pytorch/pull/65855. If I revert the `with_comms` changes from that PR (and add an additional backend parameter to the method) it seems like the decorator is working as expected.

Reply:
Should we create a separate PR for resolving this issue now? Let me know how I can help

Reply:
Hey @awaelchli @pritamdamania87 Thanks for flagging the issue! Yeah there's indeed something wrong about the behavior of `with_comms`, we should use `with_comms()` in the test case for the one that don't need to pass arguments after #65855, fixing it in https://github.com/pytorch/pytorch/pull/69493 so that we can still follow the old behavior

Reply:
Thanks @wanchaol! @awaelchli Can you apply this PR on top of https://github.com/pytorch/pytorch/pull/69493 to see if everything works as expected for the unit tests?

Reply:
Thanks a lot guys. I rebased and set the pr target to @wanchaol's branch temporarily. Tests pass locally except two of them who also fail on master. 
```
FAILED test/distributed/_sharded_tensor/test_sharded_tensor.py::TestShardedTensorChunked::test_load_state_dict_errors - RuntimeError: Process 2 exi...
FAILED test/distributed/_sharded_tensor/test_sharded_tensor.py::TestShardedTensorFromLocalShards::test_init_from_local_shards_invalid_pin_memory - 
```
which could also be just a problem only I have locally.

Reply:
@awaelchli Thanks a lot! Could you also sign the CLA as mentioned in this comment: https://github.com/pytorch/pytorch/pull/68806#issuecomment-976352471 so that we can import this PR?

Reply:
I signed the CLA when I opened the PR against master. The PR is now pointing to wanchaol/192/head, this must have reset the CLA bot. I submitted CLA once again a few hours ago, but the bot hasn't update the message. 
Should I point back to master?

Reply:
@awaelchli Ah sorry for the confusion, I meant just checking locally with https://github.com/pytorch/pytorch/pull/69493 patched to see if the unit tests work as expected. I think we should wait for https://github.com/pytorch/pytorch/pull/69493 to land, rebase this PR on top of that on master and then have this PR against master :)

Reply:
Regarding on some test failures, this issue unfortunately hided some silent failures that we couldn't capture earlier, just updated the PR and should fixed all issues.

Reply:
regarding CLA, as long as the state_dict related tests passed, I think you can re-point to master, i will try to merge #69493 as soon as i can. I do notice that `test_load_state_dict_errors` failed both on master and on my fix (as it's without `with_comms` wrapper, looks like a RPC related error). We can follow up on this separately I think.

Reply:
Sounds good. Thanks! Yes the state dict tests passed when rebased on top of #69493